### ISSUE TITLE
The qualified-id has exceeded its designated extent in the intent of certain rules

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/draft#4791]
In the current draft, [basic.lookup.qual#general-2] specifies the qualified name, where it seems that it intends to use `qualified-id` to cover such as a simple-type-specifier that has the form *nested-name-specifier<sub>opt</sub>type-name*.   In other words, they all can form **nested-name-specifier identifier** on syntax.  However, they are not the same thing on grammar.   

Furthermore, [temp.res#general-5] also intends to use qualified-id to cover a simple-type-specifier that may have the aforementioned form. The meaning of the qualified-id is vague in such use, consider a using-declarator, which has the form *nested-name-specifier unqualified-id*
````cpp
using T::identifier;
```` 
Does that mean any rule that uses the qualified-id to cover the form that is not a qualified-id on grammar can also impose on the using-declarator that has the same form? It seems a defect in the current standard.  This issue is also mentioned on [SO](https://stackoverflow.com/questions/66407140/is-it-the-intent-of-the-standard-that-any-form-nested-type-specifier-followed-b)

**Comment by jensmaurer** [comment]
First of all, a qualified name as defined in [basic.lookup.qual.general] is very much different from a _qualified-id_, and intentionally so.

Second, [temp.res.general] p5 expressly uses _qualified-id_ and in the next sentence discusses _using-declarator_; applying an "if and only if" clarification to the second sentence in p5 fixes that issue without messing with _qualified-id_.

@opensdh, any thoughts here?

**Comment by xmh0511** [comment]
Consider this example
````cpp
struct Test{
  typedef int type;
};
Test::type a = 0; //#1
````
Is the terminal name in `Test::type` called a qualified name? I think it should be a qualified name, however, we cannot find *simple-type-specifier* in the list of [basic.lookup.qual#general-2]. Presumably, the rule intends to category the `simple-type-specifier` into qualified-id, otherwise, it does not make sense.  On grammar, a `simple-type-specifier` is not a qualified-id.  @opensdh  has confirmed this issue and said it could be considered a defect, please see this [comment](https://stackoverflow.com/questions/66407140/is-it-the-intent-of-the-standard-that-any-form-nested-type-specifier-followed-b#comment117416465_66411761).

**Comment by jensmaurer** [comment]
I'm certainly not going to address any specification ambiguities of a _qualified-id_ vs. a _simple-type-specifier_ vs. a _using-declarator_ etc. in an editorial issue.

**Comment by jensmaurer** [comment]
No, `type` is the qualified name here, not `T::type`, so this would be saying something different.

I think this is the operative paragraph where we differentiate between _qualified-id_ and _simple-type-specifier_ in a dependent context. I think it's too early to talk about either here, and we should simply talk about the grammatical form _nested-name-specifier_ `::` _identifier_ and its disambiguation here.

**Comment by jensmaurer** [comment]
If we already know we have a _simple-type-specifier_ (because that's the only possibility per the grammar), we don't need temp.res.general p5 to tell us we're dealing with a type. It's only in the case where the general grammar rules would allow parsing as a _qualified-id_ that we need the type-only context rules.

**Comment by xmh0511** [comment]
> I think this is the operative paragraph where we differentiate between qualified-id and simple-type-specifier in a dependent context. I think it's too early to talk about either here, 

Violent agree! *qualified-id* can only appear as an expression, if we directly say it's a `qualified-id`, it has already meant it's an expression rather than anything others. The ambiguous case is
````cpp
typename T::type * id;
````
`T::type` could be either a qualified-id or a type-specifier. However, we cannot arbitrarily say it's just a qualified-id and try to use [temp.res#general-5] to state it's considered to denote a type. It's contradictory.  

In addition, I think the grammar *nested-name-specifier :: identifier* is not sufficient to list all possible components, such as `T::type<template-argument-list>`.

**Comment by jensmaurer** [comment]
No, `typename T::type` is exactly not ambiguous (because it's a _typename-specifier_). Without `typename`, it's ambiguous, though.

**Comment by xmh0511** [comment]
That has enforced the subject of this issue. Since the component that is in **type-only** context is one of them: *typename-specifier*, *nested-name-specifier*, *elaborated-type-specifier*, *class-or-decltype*, *type-specifier*, *decl-specifier*. How could these components be a qualified-id?

**Comment by xmh0511** [comment]
````cpp
T::type * id;
````
We could say `T::type` may be a qualified-id or a simple-type-specifier(it's ambiguous in this context). If its terminal name is in a type-only context, we say `T::type` is interpreted as a simple-type-specifier; otherwise, it's... However, if we say it's a qualified-id at the beginning, we have already designated it's just an expression, an expression can never be a type as per [expr.prim.id.qual#5].

**Comment by jensmaurer** [comment]
I think we only need to discuss the two forms "nested-name-specifier :: identifier" and "nested-name-specifier :: simple-type-id". Everything else is already unambiguous between simple-type-specifier and qualified-id.

**Comment by opensdh** [comment]
For the record, I don't have anything to add here beyond what we said in the CWG review, which was itself not a complete answer.
